src/us/kbase/test/auth2/lib/AuthenticationAnonymousIDsTest.java

```diff
@@ -0,0 +1,193 @@
+	
+	@Before
+	public void before() {
+		logEvents.clear();
+	}
+	
+	@Test
+	public void getUserNamesFromAnonymousIDsEmptyInput() throws Exception {
+		final TestMocks testauth = initTestMocks();
+		final AuthStorage storage = testauth.storageMock;
+		final Authentication auth = testauth.auth;
+		
+		final IncomingToken token = new IncomingToken("foobar");
+		
+		when(storage.getToken(token.getHashedToken())).thenReturn(
+				StoredToken.getBuilder(TokenType.LOGIN, UUID.randomUUID(), new UserName("admin"))
+						.withLifeTime(Instant.now(), Instant.now()).build());
+		when(storage.getUser(new UserName("admin"))).thenReturn(AuthUser.getBuilder(
+				new UserName("admin"), UUID.randomUUID(), new DisplayName("d"), NOW)
+				.withRole(Role.ADMIN)
+				.build());
+		
+		final Map<UUID, UserName> ret = auth.getUserNamesFromAnonymousIDs(token, set());
+		
+		assertThat("incorrect user map", ret, is(Collections.emptyMap()));
+	}
+	
+	@Test
+	public void getUserNamesFromAnonymousIDs() throws Exception {
+		// includes test of removing root user
+		// includes test that dev tokens work
+		final TestMocks testauth = initTestMocks();
+		final AuthStorage storage = testauth.storageMock;
+		final Authentication auth = testauth.auth;
+		
+		final IncomingToken token = new IncomingToken("foobar");
+		
+		final Map<UUID, UserName> expected = ImmutableMap.of(
+				UID1, new UserName("foo"), UID2, new UserName("bar"));
+
+		when(storage.getToken(token.getHashedToken())).thenReturn(
+				StoredToken.getBuilder(TokenType.DEV, UUID.randomUUID(), new UserName("admin"))
+						.withLifeTime(Instant.now(), Instant.now()).build());
+		when(storage.getUser(new UserName("admin"))).thenReturn(AuthUser.getBuilder(
+				new UserName("admin"), UUID.randomUUID(), new DisplayName("d"), NOW)
+				.withRole(Role.ADMIN)
+				.build());
+		final Map<UUID, UserName> mockret = new HashMap<>(); // needs to be mutable
+		final UUID uid3 = UUID.randomUUID();
+		mockret.put(UID1, new UserName("foo"));
+		mockret.put(UID2, new UserName("bar"));
+		mockret.put(uid3, new UserName("***ROOT***"));
+		when(storage.getUserNamesFromAnonymousIDs(set(UID2, UID1, uid3))).thenReturn(mockret);
+		
+		final Map<UUID, UserName> ret = auth.getUserNamesFromAnonymousIDs(token, set(
+				UID1, UID2, uid3));
+		
+		assertThat("incorrect user map", ret, is(expected));
+	}
+	
+	@Test
+	public void getUserNamesFromAnonymousIDsFailNulls() throws Exception {
+		final Authentication auth = initTestMocks().auth;
+		
+		final IncomingToken t = new IncomingToken("foobar");
+		final Set<UUID> u = set(UID1);
+		
+		getUserNamesFromAnonymousIDsFail(auth, null, u, new NullPointerException("token"));
+		getUserNamesFromAnonymousIDsFail(auth, t, null, new NullPointerException("anonymousIDs"));
+		getUserNamesFromAnonymousIDsFail(auth, t, set(UID1, null), new NullPointerException(
+				"Null ID in anonymousIDs"));
+	}
+	
+	@Test
+	public void getUserNamesFromAnonymousIDsExecuteStandardUserCheckingTests() throws Exception {
```

Comment:
Is this the test that checks that permissions are correct?

Reply:
It checks a bunch of stuff in the `getUser` call, including incorrect roles, no roles, no token, failing token type (which isn't relevant here), inconsistent db issues, and disabled users.

Almost every method pulls the user via that `getUser` function so it made sense to make a abstract fixture that tests all that stuff.